### PR TITLE
Optimize Cache lookup

### PR DESCRIPTION
**Pull request by fmatthew5876**

### CODE REVIEW

**Comment by fmatthew5876** [comment]
This one also makes a measurable improvement in Frozen triggers. 2% of frame time is spent in `std::map` lookups.

Without this PR
cacheold

With this PR
cachenew

**Comment by fmatthew5876** [comment]
Adding constexpr for Spec yields more small improvements.

constexprcache

**Comment by fmatthew5876** [comment]
After removing the async check.

cache-nocheck

File: src/cache.cpp [hunk 1]
```diff
@@ -262,13 +275,16 @@ namespace {
 			return LoadDummyBitmap<T>(s.directory, f);
 		}
 
+#ifndef NDEBUG
 		// Test if the file was requested asynchronously before.
 		// If not the file can't be expected to exist -> bug.
+		// This text is expensive and turned off in release builds.
```
Comment (carstene1ns) [inline src/cache.cpp]: > test

File: src/cache.cpp [hunk 2]
```diff
@@ -274,13 +275,16 @@ namespace {
...
 		FileRequestAsync* request = AsyncHandler::RequestFile(s.directory, f);
 		if (!request->IsReady()) {
-			Output::Debug("BUG: File Not Requested: %s/%s", s.directory, f.c_str());
+			Output::Error("BUG: File Not Requested: %s/%s", s.directory, f.c_str());
```
Comment (carstene1ns) [inline src/cache.cpp]: Error will currently close Player. Might be better to use Warning until #858.

File: src/cache.cpp (hunk 1 above)
Comment (fmatthew5876) [reply src/cache.cpp]: FIxed

File: src/cache.cpp (hunk 2 above)
Comment (fmatthew5876) [reply src/cache.cpp]: There are a lot of places where they just assume the cache will return non-null and then immediately start using the object. I'd rather we crash here with a big error message than get a mysterious segfault later on.

This error indicates a bug in Player code, and not game event code.

Open to suggestions though. What do you guys prefer?

Comment (Ghabry) [reply src/cache.cpp]: my original intention when writing this (with return nullptr) was to segfault right after to have a stacktrace.

Comment (fmatthew5876) [reply src/cache.cpp]: Agree let's use assert

**Comment by Ghabry** [comment]
Could you discuss the advantage of "hash key via string concat" vs "hash key as tuple"?

**Comment by fmatthew5876** [comment]
The real win here is switching from tree based `map` to hash based `unordered_map`. 

As for the difference between building a composite string vs hashing tuple elements, that would need to be profiled to get a definite answer. 

Unfortunately both idioms require string copies, but it's possible the tuple version uses shorter strings more and so we could trigger the small buffer optimization more often and avoid calls to malloc().

The standard library doesn't support arbitrary hashing of tuples, for that we would need something like `boost::hash_combine`

**Comment by fmatthew5876** [comment]
jenkins test this please

**Pull request merged**